### PR TITLE
travis-ci support

### PR DESCRIPTION
**Pull request by chipitsine**

### CODE REVIEW

**Comment by chipitsine** [comment]
there's some issue with resolving openresty.org from google cloud (have a look, it might mean serious things)

Resolving openresty.org (openresty.org)... failed: Name or service not known.

wget: unable to resolve host address ‘openresty.org’

-------------------------
however, it will work on later commits, I added cache (no worry to merge this request despite it is failing)

cache:
  directories:
  - download-cache

**Comment by chipitsine** [comment]
brrrrrr

```
[31mnot ok 222 - TEST 34: github issue \#215: Handle the posted requests in lua cosocket api (successfully resolved) - response_body - response is expected (repeated req 0, req 0)[0m
[37m[44mok 223 # skip TEST 34: github issue #215: Handle the posted requests in lua cosocket api (successfully resolved) - stap_out - tests skipped due to env TEST_NGINX_USE_STAP is not set[0m

#   Failed test 'TEST 34: github issue \#215: Handle the posted requests in lua cosocket api (successfully resolved) - response_body - response is expected (repeated req 0, req 0)'
#   at /usr/local/share/perl/5.18.2/Test/Nginx/Socket.pm line 1277.
#          got: 'failed to connect to openresty.org: openresty.org could not be resolved (2: Server failure)
# '
#     expected: 'successfully connected to xxx!
# '
```

**Comment by chipitsine** [comment]
somehow, it did compile exactly once 

https://travis-ci.org/chipitsine/lua-nginx-module/builds/130197481

we should think of performing ssl tests against localhost (instead of openresty.org and iscribblet.org), I think

**Comment by chipitsine** [comment]
maybe, there's a way to skip tests on network failures?

**Comment by chipitsine** [comment]
it seems, something wrong with dns servers 

```
$ dig @b.dnspod.com openresty.org 

; <<>> DiG 9.10.3-P4-RedHat-9.10.3-12.P4.fc23 <<>> @b.dnspod.com openresty.org
; (3 servers found)
;; global options: +cmd
;; connection timed out; no servers could be reached
[ilia@localhost ]$ 

```

**Comment by agentzh** [comment]
@chipitsine I think I've fixed the DNS server issue. But still maybe it's better to dig for those test domains at the beginning of the tests and even better, introduce a local DNS cache service, as what I did in my EC2 test cluster:

https://github.com/openresty/opsboy/blob/master/samples/ortest-ec2.ob.tt#L31

BTW, I prefer testing real domains and real certificates.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+  
+before_script:
+  - mysql -uroot -e 'create database ngx_test; grant all on ngx_test.* to "ngx_test"@"%" identified by "ngx_test"; flush privileges;'
+  
+script:
+  - tar xzf download-cache/drizzle7-2011.07.21.tar.gz && cd drizzle7-2011.07.21 
+  - ./configure --prefix=/usr --without-server > build.log 2>&1 || cat build.log
+  - sudo make libdrizzle-1.0 install-libdrizzle-1.0 > build.log 2>&1 || cat build.log
+  - cd ..
+  - cd mockeagain && make && cd ..
```
Comment (agentzh) [inline .travis.yml]: Mockeagain is not required by the default test running mode. It's only needed for the mockeagain testing mode which is apparently missing in this test configuration :)

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  - sudo apt-get install -y axel cpanminus libluajit-5.1-dev libgd-dev
+  
+install:
+  - if [ ! -d download-cache ]; then mkdir download-cache; fi
+  - if [ ! -f download-cache/drizzle7-2011.07.21.tar.gz ]; then wget -O download-cache/drizzle7-2011.07.21.tar.gz http://openresty.org/download/drizzle7-2011.07.21.tar.gz; fi
+  - wget http://ftp.cs.stanford.edu/pub/exim/pcre/pcre-8.33.tar.gz
+  - git clone https://github.com/openresty/test-nginx.git
+  - git clone https://github.com/openresty/openresty.git ../openresty
+  - git clone https://github.com/openresty/no-pool-nginx.git ../no-pool-nginx
+  - git clone https://github.com/PiotrSikora/ngx_http_auth_request_module.git ~/work/nginx/ngx_http_auth_request_module-0.2
```
Comment (agentzh) [inline .travis.yml]: I've already fixed the auth request thing. We no longer need this line :)

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+ - memcache
+ - redis-server
+    
+before_install:
+  - sudo apt-get install -y axel cpanminus libluajit-5.1-dev libgd-dev
+  
+install:
+  - if [ ! -d download-cache ]; then mkdir download-cache; fi
+  - if [ ! -f download-cache/drizzle7-2011.07.21.tar.gz ]; then wget -O download-cache/drizzle7-2011.07.21.tar.gz http://openresty.org/download/drizzle7-2011.07.21.tar.gz; fi
+  - wget http://ftp.cs.stanford.edu/pub/exim/pcre/pcre-8.33.tar.gz
```
Comment (agentzh) [inline .travis.yml]: Yeah, we need the exact 8.33 version of PCRE to get the tests pass.

**Comment by chipitsine** [comment]
I'll have a look at dns caching.
btw, "dig @127.0.0.1 openresty.org || exit 0" is not about caching, it will cache nxdomain in case of failure (and return 0)

**Comment by agentzh** [comment]
@chipitsine Well, assuming the DNS server is not timed out or failing, of course :)

**Comment by agentzh** [comment]
@chipitsine There is indeed an DNS server failing. I've removed it.

File: .travis.yml (hunk 1 above)
Comment (chipitsine) [inline .travis.yml]: pardon ?

[ilia@localhost t]$ grep -i mock *
grep: 023-rewrite: Is a directory
grep: 024-access: Is a directory
057-flush-timeout.t:    if ($ENV{LD_PRELOAD} !~ /\bmockeagain\.so\b/) {
057-flush-timeout.t:        $ENV{LD_PRELOAD} = "mockeagain.so $ENV{LD_PRELOAD}";
057-flush-timeout.t:    if ($ENV{MOCKEAGAIN} eq 'r') {
057-flush-timeout.t:        $ENV{MOCKEAGAIN} = 'rw';
057-flush-timeout.t:        $ENV{MOCKEAGAIN} = 'w';
057-flush-timeout.t:    $ENV{MOCKEAGAIN_WRITE_TIMEOUT_PATTERN} = 'hello, world';
065-tcp-socket-timeout.t:    if ($ENV{LD_PRELOAD} !~ /\bmockeagain\.so\b/) {
065-tcp-socket-timeout.t:        $ENV{LD_PRELOAD} = "mockeagain.so $ENV{LD_PRELOAD}";
065-tcp-socket-timeout.t:    if ($ENV{MOCKEAGAIN} eq 'r') {
065-tcp-socket-timeout.t:        $ENV{MOCKEAGAIN} = 'rw';
065-tcp-socket-timeout.t:        $ENV{MOCKEAGAIN} = 'w';
065-tcp-socket-timeout.t:    $ENV{MOCKEAGAIN_WRITE_TIMEOUT_PATTERN} = 'get helloworld';
117-raw-req-socket-timeout.t:    if ($ENV{LD_PRELOAD} !~ /\bmockeagain\.so\b/) {
117-raw-req-socket-timeout.t:        $ENV{LD_PRELOAD} = "mockeagain.so $ENV{LD_PRELOAD}";
117-raw-req-socket-timeout.t:    if ($ENV{MOCKEAGAIN} eq 'r') {
117-raw-req-socket-timeout.t:        $ENV{MOCKEAGAIN} = 'rw';
117-raw-req-socket-timeout.t:        $ENV{MOCKEAGAIN} = 'w';
117-raw-req-socket-timeout.t:    $ENV{MOCKEAGAIN_WRITE_TIMEOUT_PATTERN} = 'hello, world';
131-duplex-req-socket.t:    if ($ENV{LD_PRELOAD} !~ /\bmockeagain\.so\b/) {
131-duplex-req-socket.t:        $ENV{LD_PRELOAD} = "mockeagain.so $ENV{LD_PRELOAD}";
131-duplex-req-socket.t:    if ($ENV{MOCKEAGAIN} eq 'r') {
131-duplex-req-socket.t:        $ENV{MOCKEAGAIN} = 'rw';
131-duplex-req-socket.t:        $ENV{MOCKEAGAIN} = 'w';
131-duplex-req-socket.t:    $ENV{MOCKEAGAIN_WRITE_TIMEOUT_PATTERN} = 'slow';
grep: cert: Is a directory
grep: data: Is a directory
grep: lib: Is a directory
[ilia@localhost t]$

**Comment by chipitsine** [comment]
I'll think about dns caching later, not now.

File: .travis.yml (hunk 1 above)
Comment (agentzh) [inline .travis.yml]: @chipitsine Okay, my bad. Some of the test files indeed enforce the use of mockeagain :)

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,84 @@
+sudo: required
+dist: trusty
+
+os: linux
+
+language: c
+
+compiler:
+  - gcc
+#  - clang
```
Comment (agentzh) [inline .travis.yml]: Why disabling clang? Is it failing or just taking too long to build/test?

**Comment by agentzh** [comment]
@chipitsine Thank you for taking care of the DNS cache. No hurry for now :)

File: .travis.yml (hunk 4 above)
Comment (chipitsine) [inline .travis.yml]: there are issues with pcre, I'll address those issues later

Comment (agentzh) [inline .travis.yml]: @chipitsine Okay, I see. Thanks! I'll merge your current PR first. Feel free to create new PRs to address these little things :)

**Comment by agentzh** [comment]
@chipitsine Seems like we should use `prove -r t` instead of `prove -lv t` to run the test suite. The latter produces too much logs which make failures much harder to spot and also incorrectly skip test files in subdirectories like `t/023-rewrite/` and `t/024-access/`.

**Comment by ghedo** [comment]
Maybe also add a valgrind build?

**Comment by ghedo** [comment]
Oh, apparently the OpenSSL version is too old. Might make sense to clone the openssl git repo, switch to the `OpenSSL_1_0_2-stable` branch and then use `--with-openssl=...` with `ngx-build`

**Comment by chipitsine** [comment]
openssl is not too old. we still use 1.0.1 on freebsd, for example.
what would make sense is to make a matrix of different openssl builds

**Comment by ghedo** [comment]
It is too old if you want to run the SSL tests:
https://travis-ci.org/openresty/lua-nginx-module#L1268

**Comment by agentzh** [comment]
Yeah, we need at least OpenSSL 1.0.2e for the `ssl_certificate_by_lua*` feature and its tests.

**Comment by agentzh** [comment]
Already merged into master. We can address remaining issues in separate PRs :)

**Comment by agentzh** [comment]
@ghedo Adding valgrind testing mode to Travis CI might be a bit tricky since we need whitelist of expected failures for those tests sensitive to timing issues (valgrind slows the tests significantly), as with my EC2 test cluster:

http://qa.openresty.org/linux_x86_64_nginx_1_9_7__no_pool__ngx_lua_v.html

**Comment by agentzh** [comment]
@ghedo The DNS thing seems fine now. Also, I've updated the travis CI build quite a bit by introducing parallel builds and reducing `TEST_NGINX_SLEEP` settings. I've also enabled the `v2.1-agentzh` branch of OpenResty's `luajit2` github repo for the LuaJIT build.

@ghedo I've added OpenSSL 1.0.2h to the Travis CI build. Those SSL test files are now actually running :)

**Comment by agentzh** [comment]
@chipitsine I've also enfornced clang to build mockeagain in the clang builds too.

**Comment by agentzh** [comment]
@ghedo Valgrind is slow and we have so many tests in this project. Running the test suite with Valgrind can take up to 100+ min (see http://qa.openresty.org/linux_x86_64_nginx_1_9_7__no_pool__ngx_lua_v.html), which will always exceed the build time limit on travis-ci.org (50min):

https://docs.travis-ci.com/user/customizing-the-build/#Build-Timeouts

**Pull request closed without merging**